python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py

```diff
@@ -0,0 +1,204 @@
+    def size(self) -> int:
+        if self.__batches is None:
+            return 0
+        else:
+            return len(self.__batches)
+
+    def add_batches(
+        self,
+        df: Union[cudf.DataFrame, dask_cudf.DataFrame],
+        start_col_name: str,
+        batch_col_name: str,
+    ) -> None:
+        """
+        Adds batches to this BulkSampler.
+
+        Parameters
+        ----------
+        df: cudf.DataFrame or dask_cudf.DataFrame
+            Contains columns for vertex ids, batch id
+        start_col_name: str
+            Name of the column containing the start vertices
+        batch_col_name: str
+            Name of the column containing the batch ids
+
+        Returns
+        -------
+        """
+        df = df.rename(
+            columns={
+                start_col_name: self.start_col_name,
+                batch_col_name: self.batch_col_name,
+            }
+        )
+
+        if self.__batches is None:
+            self.__batches = df
+        else:
+            if isinstance(df, type(self.__batches)):
+                self.__batches = self.__batches.append(df)
+            else:
+                raise TypeError(
+                    "Provided batches must match the dataframe"
+                    " type of previous batches!"
+                )
+
+        while self.size >= self.saturation_level:
+            self.flush()
+
+    def flush(self, skip_partition_size_check=False) -> None:
+        """
+        Computes all uncomputed batches
+        """
+        min_batch_id = self.__batches[self.batch_col_name].min()
+        if isinstance(self.__batches, dask_cudf.DataFrame):
+            min_batch_id = min_batch_id.compute()
+        min_batch_id = int(min_batch_id)
+
+        partition_size = self.batches_per_partition * self.batch_size
+        partitions_per_saturation_level = self.saturation_level // partition_size
+
+        if skip_partition_size_check:
+            npartitions = int(ceil(len(self.__batches) / partition_size))
+        else:
+            npartitions = partitions_per_saturation_level
+
+        max_batch_id = min_batch_id + npartitions * self.batches_per_partition - 1
+        batch_id_filter = self.__batches[self.batch_col_name] <= max_batch_id
+
+        sample_fn = (
+            cugraph.uniform_neighbor_sample
+            if isinstance(self.__graph._plc_graph, pylibcugraph.graphs.SGGraph)
+            else cugraph.dask.uniform_neighbor_sample
+        )
+
+        # TODO semaphore check to prevent concurrent calls to uniform_neighbor_sample
+        samples = sample_fn(
+            self.__graph,
+            **self.__sample_call_args,
+            start_list=self.__batches[self.start_col_name][batch_id_filter],
+            batch_id_list=self.__batches[self.batch_col_name][batch_id_filter],
+            with_edge_properties=True,
+        )
+
+        self.__batches = self.__batches[~batch_id_filter]
+        if len(self.__batches) == 0:
+            self.__batches = None
```

Comment:
I think we will have to make this recursive. Based on testing we dont actually flush all the batches. 
```suggestion
        if len(self.__batches) == 0:
            self.__batches = None
        else:
           self.flush()
```